lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,223 @@
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(deleteRecoveryKey)
+          .then(createResponse)
+          .then(reply, reply)
+
+        function checkRecoveryKey() {
+          return db.getRecoveryKey(uid, recoveryKeyId)
+        }
+
+        function getAccount() {
+          return db.account(uid)
+            .then((result) => account = result)
+        }
+
+        // Password reset using a recovery key behaves similarly to a password change. The client should have
+        // retrieved their original `kB` and rewrapped the kB.
+        function passwordReset() {
+          return random.hex(32)
+            .then(hex => {
+              authSalt = hex
+              password = new Password(authPW, authSalt, verifierVersion)
+              return password.verifyHash()
+            })
+            .then((hash) => {
+              verifyHash = hash
+              return password.wrap(wrapKb)
+            })
+            .then((wrapWrapKb) => {
+              return db.resetAccount(
+                {uid},
+                {
+                  verifyHash: verifyHash,
+                  authSalt: authSalt,
+                  wrapWrapKb: wrapWrapKb,
+                  verifierVersion: password.version
+                }
+              )
+            })
+        }
+
+        function createSessionToken() {
+          const {
+            browser: uaBrowser,
+            browserVersion: uaBrowserVersion,
+            os: uaOS,
+            osVersion: uaOSVersion,
+            deviceType: uaDeviceType,
+            formFactor: uaFormFactor
+          } = request.app.ua
+
+          const sessionTokenOptions = {
+            uid: account.uid,
+            email: account.primaryEmail.email,
+            emailCode: account.primaryEmail.emailCode,
+            emailVerified: account.primaryEmail.isVerified,
+            verifierSetAt: account.verifierSetAt,
+            uaBrowser,
+            uaBrowserVersion,
+            uaOS,
+            uaOSVersion,
+            uaDeviceType,
+            uaFormFactor
+          }
+
+          return db.createSessionToken(sessionTokenOptions)
+            .then((result) => sessionToken = result)
+        }
+
+        function createKeyFetchToken() {
+          if (requestHelper.wantsKeys(request)) {
+            return db.createKeyFetchToken({
+              uid: account.uid,
+              kA: account.kA,
+              wrapKb: wrapKb,
+              emailVerified: account.primaryEmail.isVerified
+            })
+              .then((result) => keyFetchToken = result)
+          }
+        }
+
+        function deleteRecoveryKey() {
+          return db.deleteRecoveryKey(uid, recoveryKeyId)
```

Comment:
Is it possible for the user to have multiple recovery keys?  If so, is it appropriate for those additional recovery keys to persist after an account reset, or should the act of resetting cause all the keys to be invalidated?

Reply:
It is currently not possible for a user to have multiple recovery keys. Although, for phase 2 we can drop this requirement and delete all recovery keys.

Reply:
👍 ; To be clear, I genuinely don't know whether it's better to delete just the one that was used, or to delete all of them and require the user to re-establish them all, can can see arguments for both sides.